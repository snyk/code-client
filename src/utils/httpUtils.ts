import { ORG_ID_REGEXP } from '../constants';
import { JsonApiError } from '../interfaces/json-api';
import { ResultError } from '../http';

export function getURL(baseURL: string, path: string, orgId?: string): string {
  if (routeToGateway(baseURL)) {
    if (!isValidOrg(orgId)) {
      throw new Error('A valid Org id is required for this operation');
    }
    return `${baseURL}/hidden/orgs/${orgId}/code${path}`;
  }
  return `${baseURL}${path}`;
}

function routeToGateway(baseURL: string): boolean {
  return baseURL.includes('snykgov.io');
}

function isValidOrg(orgId?: string): boolean {
  return orgId !== undefined && ORG_ID_REGEXP.test(orgId);
}

export function isJsonApiErrors(input: unknown): input is JsonApiError[] {
  if (!Array.isArray(input) || input.length < 1) {
    return false;
  }

  for (const element of input) {
    if (
      typeof element !== 'object' ||
      !('status' in element) ||
      !('code' in element) ||
      !('title' in element) ||
      !('detail' in element)
    ) {
      return false;
    }
  }

  return true;
}

export function generateErrorWithDetail<E>(error: JsonApiError, statusCode: number, apiName: string): ResultError<E> {
  const errorLink = error.links?.about;
  const detail = `${error.title}${error.detail ? `: ${error.detail}` : ''}${
    errorLink ? ` (more info: ${errorLink})` : ``
  }`;
  const statusText = error.title;
  return {
    type: 'error',
    error: {
      apiName,
      statusCode: statusCode as unknown as E,
      statusText,
      detail,
    },
  };
}
