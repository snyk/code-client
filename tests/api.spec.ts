import {
  AnalysisStatus,
  checkBundle,
  checkSession,
  createBundle,
  createGitBundle,
  extendBundle,
  getAnalysis,
  getFilters,
  getIpFamily,
  reportError,
  reportEvent,
  startSession,
  uploadFiles,
} from '../src/http';
import { fromEntries } from '../src/lib/utils';
import { authHost, baseURL, sessionToken, TEST_TIMEOUT } from './constants/base';
import { bundleFiles, bundleFilesFull } from './constants/sample';

const fakeBundleId = 'c58d69bd4fd65c45b1112bd7b45f028e614d443fc123901fd1aba15856c13c27';
let fakeBundleIdFull = '';
const realBundleId = 'c3a31c6c503b76ce51e2f8f7db9aa8f26715e6367124ce6e2f419087cad011b0';
let realBundleIdFull = '';
const extendedBundleId = '587a6bcb0095606ad57ccc7bb7ac6401475ce4181c13f7136491a16df06544f1';

const fakeMissingFiles = [
  `/AnnotatorTest.cpp`,
  `/GitHubAccessTokenScrambler12.java`,
  `/app.js`,
  `/db.js`,
  `/main.js`,
  // TODO: This should be ignored
  `/not/ignored/this_should_be_ignored.jsx`,
  `/not/ignored/this_should_not_be_ignored.java`,
  `/routes/index.js`,
  `/routes/sharks.js`,
];

const reportTelemetryRequest = {
  baseURL,
  sessionToken,
  bundleId: fakeBundleId,
  source: 'testSource',
  type: 'testType',
  message: 'testMessage',
  path: '/test/path',
  data: {
    foo: 'bar',
    bar: ['fo', 'foo', 'fooo'],
  },
};

describe('Requests to public API', () => {
  it('gets filters successfully', async () => {
    const response = await getFilters(baseURL, '');
    expect(response.type).toEqual('success');
    if (response.type === 'error') return;
    expect(new Set(response.value.configFiles)).toEqual(new Set(['.dcignore', '.gitignore']));
    expect(new Set(response.value.extensions)).toEqual(
      new Set([
        '.CS',
        '.Cs',
        '.c',
        '.cc',
        '.cpp',
        '.cs',
        '.cxx',
        '.es',
        '.es6',
        '.h',
        '.hpp',
        '.htm',
        '.html',
        '.hxx',
        '.java',
        '.js',
        '.jsx',
        '.php',
        '.py',
        '.ts',
        '.tsx',
        '.vue',
        '.ASPX',
        '.Aspx',
        '.aspx',
        '.ejs',
      ]),
    );
  });

  it('reports error successfully', async () => {
    const response = await reportError(reportTelemetryRequest);
    expect(response.type).toEqual('success');
  });

  it('reports event successfully', async () => {
    const response = await reportEvent(reportTelemetryRequest);
    expect(response.type).toEqual('success');
  });

  it('starts session successfully', async () => {
    // const ipFamily = await getIpFamily(authHost);

    const startSessionResponse = startSession({
      source: 'atom',
      authHost,
    });
    expect(startSessionResponse.loginURL).toMatch(/.*\/login\?token=.*&utm_source=atom/);
    const draftToken = startSessionResponse.draftToken;
    // This token is just a draft and not ready to be used permanently

    const checkSessionResponse = await checkSession({ authHost, draftToken, ipFamily: 4 });

    expect(checkSessionResponse.type).toEqual('success');
    if (checkSessionResponse.type == 'error') return;
    expect(checkSessionResponse.value).toEqual('');
  });

  it('checks session unsuccessfully', async () => {
    const response = await checkSession({
      authHost,
      draftToken: 'dummy-token',
    });
    expect(response.type).toEqual('success');
    if (response.type === 'error') return;
    expect(response.value).toEqual('');
  });

  // TODO: find a way to test successfull workflow automatically
  // it('checks session successfully', async () => {
  //   const response = await checkSession({
  //     authHost,
  //     sessionToken,
  //   });
  //   expect(response.type).toEqual('success');
  //   if (response.type === 'error') return;
  //   expect(response.value).toEqual(true);
  // });

  it(
    'creates bundle successfully',
    async () => {
      const files = fromEntries([...(await bundleFiles).entries()].map(([i, d]) => [d.bundlePath, `${i}`]));

      const response = await createBundle({
        baseURL,
        sessionToken,
        files,
        source: 'atom',
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.bundleId).toContain(fakeBundleId);
      fakeBundleIdFull = response.value.bundleId;
      expect(response.value.missingFiles).toEqual(fakeMissingFiles);
    },
    TEST_TIMEOUT,
  );

  it(
    'checks bundle successfully',
    async () => {
      const response = await checkBundle({
        baseURL,
        sessionToken,
        bundleId: fakeBundleIdFull,
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.bundleId).toEqual(fakeBundleIdFull);
      expect(response.value.missingFiles).toEqual(fakeMissingFiles);
    },
    TEST_TIMEOUT,
  );

  it(
    'checks expired bundle successfully',
    async () => {
      const response = await checkBundle({
        baseURL,
        sessionToken,
        bundleId: 'mock-expired-bundle-id',
      });
      expect(response.type).toEqual('error');
      // dummy to cheat typescript compiler
      if (response.type == 'success') return;
      expect(response.error.statusCode).toEqual(404);
      expect(response.error.statusText).toEqual('Uploaded bundle has expired');
    },
    TEST_TIMEOUT,
  );

  it(
    'request analysis with missing files',
    async () => {
      let response;
      do {
        response = await getAnalysis({
          baseURL,
          sessionToken,
          bundleId: fakeBundleIdFull,
          severity: 1,
          source: 'atom',
        });
      } while (response.type === 'success');

      expect(response.type).toEqual('error');
      expect(response.error).toEqual({
        apiName: 'getAnalysis',
        statusCode: 404,
        statusText: 'Not found',
      });
    },
    TEST_TIMEOUT,
  );

  it(
    'extends bundle successfully',
    async () => {
      const response = await extendBundle({
        baseURL,
        sessionToken,
        bundleId: fakeBundleIdFull,
        files: {
          [`/new.js`]: 'new123',
        },
        removedFiles: [
          `/AnnotatorTest.cpp`,
          `/app.js`,
          `/GitHubAccessTokenScrambler12.java`,
          `/db.js`,
          `/main.js`,
          `/not/ignored/this_should_be_ignored.jsx`,
          `/not/ignored/this_should_not_be_ignored.java`,
          `/routes/index.js`,
          `/routes/sharks.js`,
        ],
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.bundleId).toContain(extendedBundleId);
      expect(response.value.missingFiles).toEqual([`/new.js`]);
    },
    TEST_TIMEOUT,
  );

  it(
    'extends expired bundle successfully',
    async () => {
      const response = await extendBundle({
        baseURL,
        sessionToken,
        bundleId: 'wrong-bundle-id',
        files: {
          [`/new2.js`]: 'new1234',
        },
      });

      expect(response.type).toEqual('error');
      // dummy to cheat typescript compiler
      if (response.type == 'success') return;

      expect(response.error.statusCode).toEqual(404);
      expect(response.error.statusText).toEqual('Parent bundle has expired');
    },
    TEST_TIMEOUT,
  );

  it(
    'uploads fake files to fake bundle',
    async () => {
      const response = await uploadFiles({
        baseURL,
        sessionToken,
        bundleId: fakeBundleIdFull,
        content: [
          {
            fileHash: 'df',
            fileContent: 'const module = new Module();',
          },
          {
            fileHash: 'sdfs',
            fileContent: 'const App = new App();',
          },
        ],
      });
      expect(response.type).toEqual('error');
      if (response.type === 'success') return;
      expect(response.error).toEqual({
        apiName: 'uploadFiles',
        statusCode: 400,
        statusText:
          'Invalid request, attempted to extend a git bundle, or ended up with an empty bundle after the extension',
      });
    },
    TEST_TIMEOUT,
  );

  it('test successful workflow', async () => {
      // Create a bundle first
      const files = fromEntries((await bundleFilesFull).map(d => [d.bundlePath, d.hash]));

      const bundleResponse = await createBundle({
        baseURL,
        sessionToken,
        files,
        source: 'atom',
      });
      expect(bundleResponse.type).toEqual('success');
      if (bundleResponse.type === 'error') return;
      expect(bundleResponse.value.bundleId).toContain(realBundleId);
      realBundleIdFull = bundleResponse.value.bundleId;

      const content = (await bundleFilesFull).map(d => {
        return {
          fileHash: d.hash,
          fileContent: d.content || '',
        };
      });

      // Upload files
      const uploadResponse = await uploadFiles({
        baseURL,
        sessionToken,
        bundleId: realBundleIdFull,
        content: content,
      });

      expect(uploadResponse.type).toEqual('success');
      if (uploadResponse.type === 'error') return;
      expect(uploadResponse.value).toEqual(true);

      // Check missing files
      const checkResponse = await checkBundle({
        baseURL,
        sessionToken,
        bundleId: realBundleIdFull,
      });
      expect(checkResponse.type).toEqual('success');
      if (checkResponse.type === 'error') return;
      expect(checkResponse.value.bundleId).toEqual(realBundleIdFull);
      expect(checkResponse.value.missingFiles).toEqual([]);

      // Get analysis results
      let response = await getAnalysis({
        baseURL,
        sessionToken,
        bundleId: realBundleIdFull,
        severity: 1,
        source: 'atom',
      });
      expect(response.type).toEqual('success');
      if (response.type === 'error') return;
      expect(response.value.status !== AnalysisStatus.failed).toBeTruthy();

      if (response.value.status === AnalysisStatus.done) {
        expect(Object.keys(response.value.analysisResults.suggestions).length).toEqual(8);
        const suggestion = Object.values(response.value.analysisResults.suggestions).find(s => s.id === 'javascript%2Fdc_interfile_project%2FDisablePoweredBy')!;
        expect(Object.keys(suggestion)).toEqual([
          'id',
          'rule',
          'message',
          'severity',
          'lead_url',
          'leadURL',
          'categories',
          'tags',
          'title',
          'cwe',
          'text',
          'repoDatasetSize',
          'exampleCommitDescriptions',
          'exampleCommitFixes',
        ]);
        expect(suggestion.leadURL).toEqual(
          'http://expressjs.com/en/advanced/best-practice-security.html#at-a-minimum-disable-x-powered-by-header',
        );
        expect(suggestion.repoDatasetSize).toEqual(835);
        expect(suggestion.exampleCommitDescriptions).toEqual([
          'Test without express',
          'secure the api with helmet',
          'Fix some security vulnerabilities, refactor the error constants a little',
        ]);
        expect(suggestion.exampleCommitFixes.length).toEqual(3);
        expect(suggestion.message).toEqual(
          'Disable X-Powered-By header for your Express app (consider using Helmet middleware), because it exposes information about the used framework to potential attackers.',
        );
        expect(suggestion.rule).toEqual('DisablePoweredBy');
        expect(suggestion.severity).toEqual(2);

        expect(suggestion.tags).toEqual(['maintenance', 'express', 'helmet', 'security']);
        expect(Object.keys(response.value.analysisResults.files).length).toEqual(5);
        expect(response.value.analysisResults.timing.analysis).toBeGreaterThanOrEqual(
          response.value.analysisResults.timing.fetchingCode,
        );
        const filePath = `/GitHubAccessTokenScrambler12.java`;
        expect(response.value.analysisResults.files[filePath]).toMatchSnapshot();
        expect(response.value.analysisResults.timing.queue).toBeGreaterThanOrEqual(0);
        expect(new Set(response.value.analysisResults.coverage)).toEqual(
          new Set([
            {
              files: 2,
              isSupported: true,
              lang: 'Java',
            },
            {
              files: 1,
              isSupported: true,
              lang: 'C++ (beta)',
            },
            {
              files: 5,
              isSupported: true,
              lang: 'JavaScript',
            },
            {
              files: 1,
              isSupported: true,
              lang: 'JSX',
            },
          ]),
        );
      }

      // Get analysis results limited to 1 file
      do {
        response = await getAnalysis({
          baseURL,
          sessionToken,
          bundleId: realBundleIdFull,
          severity: 1,
          limitToFiles: [`/GitHubAccessTokenScrambler12.java`],
          source: 'atom',
        });

        expect(response.type).toEqual('success');
        if (response.type === 'error') return;
        expect(response.value.status !== AnalysisStatus.failed).toBeTruthy();
      } while (response.value.status !== AnalysisStatus.done);

      expect(Object.keys(response.value.analysisResults.suggestions).length).toEqual(4);
      expect(new Set(Object.keys(response.value.analysisResults.files))).toEqual(
        new Set(['/GitHubAccessTokenScrambler12.java', '/not/ignored/this_should_not_be_ignored.java']),
      );

      // Get analysis results with severity 3
      do {
        response = await getAnalysis({
          baseURL,
          sessionToken,
          bundleId: realBundleIdFull,
          severity: 3,
          source: 'atom',
        });
        expect(response.type).toEqual('success');
        if (response.type === 'error') return;
        expect(response.value.status !== AnalysisStatus.failed).toBeTruthy();
      } while (response.value.status !== AnalysisStatus.done);

      expect(Object.keys(response.value.analysisResults.suggestions).length).toEqual(2);
      expect(new Set(Object.keys(response.value.analysisResults.files))).toEqual(
        new Set(['/GitHubAccessTokenScrambler12.java', '/not/ignored/this_should_not_be_ignored.java']),
      );
    },
    TEST_TIMEOUT,
  );

});
