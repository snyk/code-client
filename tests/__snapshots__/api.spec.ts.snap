// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Requests to public API test successful workflow 1`] = `
Object {
  "0": Array [
    Object {
      "cols": Array [
        39,
        53,
      ],
      "fingerprints": Array [
        Object {
          "fingerprint": "369aebd49780d3608e1ed3c48a88b50d531409e64105fa20b0e9b751f066f87b",
          "version": 0,
        },
        Object {
          "fingerprint": "19689298.883ce888.fa1beb9c.44ab356c.f81b5a8d.f7edcf51.0d475b83.41ec51de.19689298.601f8825.fa1beb9c.de031890.f81b5a8d.ce109560.0d475b83.41ec51de",
          "version": 1,
        },
      ],
      "markers": Array [
        Object {
          "msg": Array [
            16,
            21,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                40,
                57,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                10,
                10,
              ],
            },
          ],
        },
        Object {
          "msg": Array [
            36,
            62,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                39,
                53,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                15,
                15,
              ],
            },
          ],
        },
      ],
      "rows": Array [
        15,
        15,
      ],
    },
  ],
  "1": Array [
    Object {
      "cols": Array [
        42,
        54,
      ],
      "fingerprints": Array [
        Object {
          "fingerprint": "e7c159b1f0c8449a86abb7a07127d4f593e958f6a3912221ba564e18d8f7e43b",
          "version": 0,
        },
        Object {
          "fingerprint": "a75b04d3.0b7f48d1.7a59bef1.44ab356c.a698423b.f7edcf51.c23c997b.41ec51de.a75b04d3.0b7f48d1.3c7d90aa.de031890.8b2d3351.ce109560.c23c997b.41ec51de",
          "version": 1,
        },
      ],
      "markers": Array [
        Object {
          "msg": Array [
            16,
            21,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                34,
                51,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                11,
                11,
              ],
            },
          ],
        },
        Object {
          "msg": Array [
            36,
            45,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                42,
                54,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                16,
                16,
              ],
            },
          ],
        },
      ],
      "rows": Array [
        16,
        16,
      ],
    },
  ],
  "2": Array [
    Object {
      "cols": Array [
        37,
        47,
      ],
      "fingerprints": Array [
        Object {
          "fingerprint": "b964e2b3c7ac4c5b425b9419abe756f566b7efcc16786e8036d44e33f109d545",
          "version": 0,
        },
        Object {
          "fingerprint": "19689298.883ce888.fdf08d99.d7919eeb.74c2d841.f7edcf51.5721ebe2.54d46e25.19689298.883ce888.fdf08d99.d7919eeb.74c2d841.f7edcf51.5721ebe2.54d46e25",
          "version": 1,
        },
      ],
      "markers": Array [
        Object {
          "msg": Array [
            4,
            6,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                49,
                70,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                18,
                18,
              ],
            },
          ],
        },
        Object {
          "msg": Array [
            21,
            51,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                37,
                47,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                18,
                18,
              ],
            },
          ],
        },
      ],
      "rows": Array [
        18,
        18,
      ],
    },
  ],
  "3": Array [
    Object {
      "cols": Array [
        14,
        15,
      ],
      "fingerprints": Array [
        Object {
          "fingerprint": "5af7ed14642e35fa42e5dcbec421163e53f6f8a9b540f1950c0cc520000d85e4",
          "version": 0,
        },
        Object {
          "fingerprint": "c2e08f55.e9fd9d38.d1699128.f482feb1.a485742e.730a5695.62adb608.e9000d02.c2e08f55.ef94efde.cd271e66.e22980a8.d31a8364.730a5695.bae646e8.54d46e25",
          "version": 1,
        },
      ],
      "markers": Array [
        Object {
          "msg": Array [
            28,
            46,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                19,
                27,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                23,
                23,
              ],
            },
          ],
        },
        Object {
          "msg": Array [
            80,
            94,
          ],
          "pos": Array [
            Object {
              "cols": Array [
                17,
                31,
              ],
              "file": "/GitHubAccessTokenScrambler12.java",
              "rows": Array [
                24,
                24,
              ],
            },
          ],
        },
      ],
      "rows": Array [
        24,
        24,
      ],
    },
  ],
}
`;
